### ISSUE TITLE
The listening socket is closed twice after a connection has been accepted

### ISSUE THREAD

**Issue by m-lw** [slime/slime#390]
When a connection has been accepted, `swank::accept-connections` closes the listening socket fd and then sends `:stop-server` to the sentinel which closes the socket again.  In rare cases, another thread can open the same fd for other purposes before the second close, leading to errors.

Please consider the patch below.

```
diff --git a/swank.lisp b/swank.lisp
index 2cdaef02..1f33fd43 100644
--- a/swank.lisp
+++ b/swank.lisp
@@ -788,11 +788,9 @@ first."
                      (accept-connection socket :external-format nil
                                                :buffering t)
                   (unless dont-close
-                    (close-socket socket)))))
+                    (send-to-sentinel `(:stop-server :socket ,socket))))))
     (authenticate-client client)
-    (serve-requests (make-connection socket client style))
-    (unless dont-close
-      (send-to-sentinel `(:stop-server :socket ,socket)))))
+    (serve-requests (make-connection socket client style))))
 
 (defun authenticate-client (stream)
   (let ((secret (slime-secret)))
```

**Comment by luismbo** [comment]
@m-lw this change makes sense to me since (a) the sentinel's purpose is to handle concurrency in operations like `:stop-server` so it doesn't make sense to have multiple threads doing a `close-socket` concurrently since it's not thread-safe and (b) seems bad style to close the same socket twice.

I don't understand your claim that another can open the same fd. Surely `close-socket` drops the fd, doesn't it? SBCL's `sb-bsd-sockets:socket-close` seems to.

**Comment by m-lw** [comment]
@luismbo thanks for picking this up.

The first call to `close-socket` does indeed drop the fd, but that allows the OS to reuse the fd immediately for some other purpose.  Making a second call to `close-socket` will then drop the fd again, even though it is not associated with slime's socket.  In the case we found, the fd had been reused as a file handle by `dlopen` in another thread before being closed for the second time, which caused a SIGSEGV after `mmap` failed with this fd.

**Comment by luismbo** [comment]
@m-lw what Lisp is that? If I'm reading SBCL's `socket-close` right, it definitely drops the FD and will never try to close it again. See: https://github.com/sbcl/sbcl/blob/622c9daf9bb41ef9ad4b8a063c62c4baf59a1c1a/contrib/sb-bsd-sockets/sockets.lisp#L249

**Comment by luismbo** [comment]
BTW, do you have a real name that I can credit in the commit message? :-)

**Comment by m-lw** [comment]
@luismbo I am Martin Simmons at LispWorks.  In our implementation, the socket is just the fd, so it can't detect double-closing.

**Comment by luismbo** [comment]
Oh, hi Martin. :-)

**Comment by luismbo** [comment]
Alas, this fails the `disconnect-one-connection` test in `slime-tests.el`. It's not obvious to me why. I'll have to look into it later.

**Comment by luismbo** [comment]
So the problem detected by the `disconnect-one-connection` test was that sending `:stop-server` to the sentinel _before_ creating a new connection would cause the sentinel to see a state with no server and no connections thus making it decide to exit.

I pondered replacing the sentinel mechanism with a lock, twisting `accepting-connections` to make sure it stopped the server before serving requests but after creating the new connection, or even adding a `:dont-close` argument to the `:stop-server` message. In the end, I went with just stopping the server at the very end, as before. Hopefully leaving the socket open for a while longer (in non multithreaded backends) isn't problematic.

Hope this fixes your issue.
